packages/account-tree-controller/src/backup-and-sync/service/index.ts

```diff
@@ -0,0 +1,363 @@
+            }
+
+            // Create local groups for each group from user storage if they do not exist
+            // This will ensure that we have all groups available locally before syncing metadata
+            await createLocalGroupsFromUserStorage(
+              this.#context,
+              groupsFromUserStorage,
+              entropySourceId,
+              walletProfileId,
+            );
+
+            // Sync group metadata bidirectionally
+            await syncGroupsMetadata(
+              this.#context,
+              wallet,
+              groupsFromUserStorage,
+              entropySourceId,
+              walletProfileId,
+            );
+          } catch (error) {
+            this.#context.contextualLogger.error(
+              `Error syncing wallet ${wallet.id}:`,
+              error instanceof Error ? error.message : String(error),
+            );
+
+            // Attempt to rollback state changes for this wallet
+            try {
+              if (!stateSnapshot) {
+                throw new Error(
+                  `State snapshot is missing for wallet ${wallet.id}`,
+                );
+              }
+              restoreStateFromSnapshot(this.#context, stateSnapshot);
+              this.#context.contextualLogger.log(
+                `Rolled back state changes for wallet ${wallet.id}`,
+              );
+            } catch (rollbackError) {
+              this.#context.contextualLogger.error(
+                `Failed to rollback state for wallet ${wallet.id}:`,
+                rollbackError instanceof Error
+                  ? rollbackError.message
+                  : String(rollbackError),
+              );
+            }
+
+            // Continue with next wallet instead of failing the entire sync
+            continue;
+          }
+        }
+      } catch (error) {
+        this.#context.contextualLogger.error(
+          'Error during multichain account syncing:',
+          error,
+        );
+        throw error;
+      } finally {
+        this.#context.controllerStateUpdateFn(
+          (state: AccountTreeControllerState) => {
+            state.isAccountTreeSyncingInProgress = false;
+            state.hasAccountTreeSyncingSyncedAtLeastOnce = true;
+          },
+        );
+      }
+    };
+
+    // Execute the big sync function with tracing
+    await this.#context.traceFn(
+      {
+        name: TraceName.AccountSyncFull,
+      },
+      bigSyncFn,
+    );
+  }
+
+  /**
+   * Performs a single wallet's bidirectional metadata sync with user storage.
+   *
+   * @param walletId - The wallet ID to sync.
+   */
+  async #performSingleWalletSync(walletId: AccountWalletId): Promise<void> {
+    try {
+      const wallet =
+        this.#context.controller.state.accountTree.wallets[walletId];
+      if (!wallet || wallet.type !== AccountWalletType.Entropy) {
+        return; // Only sync entropy wallets
+      }
```

Comment:
Very nit: We have this pattern twice, maybe we could wrap this in a `#getEntropyWallet(walletId): AccountWalletEntropyObject` so we just have to do `if (!wallet)` afterward? (Thus, enforcing the wallet type in one place).